src/mlpack/core/data/dict_encoding_impl.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * @file dict_encoding_impl.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "dict_encoding.hpp"
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+namespace mlpack {
+namespace data {
+
+/**
+ * Dictionary Encoding
+ * here we simply assign a word (or a character) to a numeric index
+ * and treat the dataset as categorical.
```

Comment:
It looks like the output matrix here is padded such that all the inputs are the same length.  But I think that this may not always be what the user wants; technically if I pass in `std::vector<std::string>` then what I should get back is a `std::vector<arma::Row<size_t>>`.  So there is some complexity to the interface here that we have to be really clear about, and think how a user might use it.

You might consider starting this with a really simple function that, given a string only, will map only that string to an output.  So, i.e., this:

```
template<typename OutputColType = arma::rowvec>
void Encode(const std::string& input, std::unordered_map<std::string, size_t>& dictionary, OutputColType&& output);
```

and I used a template parameter for `OutputColType` because users may want to do something like this...

```
Encode(string, dictionary, myMatrix.col(3));
```

or they might want to put the output into other types too.

Once you build that, then I think building other more complex functionality can be done by building on top of the simple `Encode()` function.